### DISCUSSION TITLE
Minimum time defined in the experimental protocol

### DISCUSSION THREAD

**Discussion by optibikash** [pybamm-team/PyBaMM#2013] [Q&A]
Hi everyone!
The question is out of curiosity while experimenting with PyBaMM. In the pybamm.Experiment class, we define the time for which we wish to charge, discharge, or rest the battery. Is there a limitation to the definition of time? 
Taking from the example script:
           `  "Discharge at C/5 for 10 hours or until 3.3 V",
            "Rest for 1 hour",
            "Charge at 1 A until 4.1 V",
            "Hold at 4.1 V until 10 mA",
            "Rest for 1 hour", `

Is it possible to define the time in milliseconds or microseconds? Do we have a limit for the definition of time?

**Comment by valentinsulzer** [comment]
Seconds is the smallest. But `1e-3 s` should work (open an issue if it doesn't)

**Reply by optibikash** [reply]
Thanks, Tino. I did try it but it takes a lot of time to perform simulation. The following challenges are encountered:
1. The simulation is very slow.
2. It was not possible to simulate the model in a laptop with 16 GB RAM. I tried google collab and found the failure of simulation due to lack of memory.
3. Later I tried to run the simulation in a supercomputing node of my institute server. I was able to run it and get results, although it takes a lot of time, the saving of data was not done in csv file. Hence I tried to save a pickle file which was successful but the size of the file is more than 30 GB.
4. Further, the operator of the supercomputing server also stopped the simulation later when I increased the number of cycles citing higher memory access.

I think there could be a possibility to reduce the computation time. Also, is there a way to solve the model for a particular set of parameters only? I know we can save a particular set of parameters but is there a way to solve for a particular set. 
For example, the following code simulates the model:

`
sim = pybamm.Simulation(
    model, parameter_values=param, solver=pybamm.CasadiSolver(mode="fast")
)
sim.solve()
`
In this, is there a way to define something to decrease the simulation time, reduce the requirement of RAM, solve for a particular parameter only, and save values only at some defined time intervals or some samples of data at particular intervals?

**Reply by valentinsulzer** [reply]
The answer to these questions depends on: why are you wanting to simulate every millisecond?

**Reply by optibikash** [reply]
Thanks, @tinosulzer for the response and my apologies for the late responses. I was travelling and couldn't track the response. 
I want to simulate the pulse charging in the PyBamm. I have my own built testbed and I could perform the pulse charging experiments as well. I wish to relate the experimental findings with the electrochemical models. I wished not to increase the frequency of the pulse above 10kHz but I found a thesis (from Texas A&M University) where similar experiments although not very detailed analyses are done at 100 kHz as well. 
However, simulating at a 10 kHz pulse is also challenging. It would a nice to get your insights into implementing the idea in PyBamm.

**Reply by optibikash** [reply]
Haven't heard back from you @tinosulzer . Can you suggest anything?

**Reply by valentinsulzer** [reply]
Please provide a complete code. The one line you gave is very generic and not enough information

**Reply by optibikash** [reply]
Sure @tinosulzer .This is one of the basic code 
`
#Pulse Charging#

import csv
import pybamm
import numpy as np
import datetime

model = pybamm.lithium_ion.DFN(options = {"thermal": "lumped", "SEI":"reaction limited", "SEI porosity change" : "true", "lithium plating": "irreversible"})
parameter_values = pybamm.ParameterValues(chemistry=pybamm.parameter_sets.Chen2020_plating)
param = parameter_values
t1 = datetime.datetime.now()
f = open('completed.txt', 'w+')
f.write('Program execution started')
cycle = ["Discharge at C/30 until 2.8V"] + [
    "Charge at 2C for 0.001 seconds", #Increase charge period
    "Rest for 0.001 seconds"] * 225000 + ["Rest for 2 hours",
        "Charge at C/30 until 4.2V", #Increase charge period
        "Rest for 2 hours",
        "Discharge at C/30 until 2.8V"]

experiment1 = pybamm.Experiment(cycle)

sim = pybamm.Simulation(model, experiment=experiment1, parameter_values=param)

sim.solve(solver=pybamm.CasadiSolver(mode="safe", dt_max=1000))
t2 = datetime.datetime.now()

sol = sim.solution

msg = "Program execution was completed at "+str(t1)+".\nTotal time elapsed\t"+str(t2-t1)+"."
f.write(msg)
f.close()
sol.save_data("Efficiency_500hz.csv", ["Time [s]", "Terminal voltage [V]" , "Current [A]","X-averaged cell temperature [K]"], to_format="csv")
sol.save_data("Efficiency_500hz.json", ["Time [s]", "Terminal voltage [V]" , "Current [A]","X-averaged cell temperature [K]"], to_format="json")
`

**Reply by koomarmdey** [reply]
Hey, I was looking into a similar project to simulate pulse charging using PyBaMM. Is it possible to this by defining small timesteps in my experiment? or is there any better way?

 Any suggestions would be appreciated. Thanks

**Reply by optibikash** [reply]
@koomarmdey 
I am not sure which version you are using, but I was using quite an old one. In the newer versions, I have not tried implementing pulse charging. You should have taken a look into the old comments, and you might have got the response. Quoting again what information was provided to me by @tinosulzer  "Seconds is the smallest. But 1e-3 s should work (open an issue if it doesn't)."
Maybe with newer versions, it would have much smaller time steps compared to what it was 2 years back. The type of solver you select also plays a crucial role in the defining time steps. Since my code is already pasted in the previous comments, you know what I attempted and how. I hope this helps!
